torch/csrc/distributed/rpc/testing/faulty_tensorpipe_agent.cpp

```diff
@@ -0,0 +1,151 @@
+  for (const auto& msgString : messagesToFail) {
+    messageTypesToFail.push_back(messageStringToType(msgString));
+  }
+  return messageTypesToFail;
+}
+
+std::unordered_map<MessageType, float, std::hash<int>> FaultyTensorPipeAgent::
+    parseMessagesToDelay(const std::unordered_map<std::string, float>&
+                             messageTypesToDelay) const {
+  std::unordered_map<MessageType, float, std::hash<int>> delayMessages;
+  for (const auto& messagePair : messageTypesToDelay) {
+    float delay = messagePair.second;
+    TORCH_CHECK(
+        delay >= 0,
+        "Delays passed to FaultyTensorPipeAgent must be non-negative.")
+    delayMessages.insert({messageStringToType(messagePair.first), delay});
+  }
+  return delayMessages;
+}
+
+c10::intrusive_ptr<JitFuture> FaultyTensorPipeAgent::send(
+    const WorkerInfo& to,
+    c10::intrusive_ptr<Message> message,
+    const float rpcTimeoutSeconds,
+    const std::unordered_map<c10::Device, c10::Device>& /* unused */) {
+  // We only fail control messages that have been specified by the test case.
+  // For all other messages, we just send them without any failures.
+  if (!shouldFailMessage(message->type())) {
+    return TensorPipeAgent::send(to, std::move(message), rpcTimeoutSeconds);
+  }
+
+  // This send function checks the failMessageCountMap_ to check whether
+  // we must fail the next send. If the send must be failed, we set an error
+  // on the returned future immediately and increment the counter in the map,
+  // otherwise we just call the TensorPipeAgent send.
+  const auto key = fromVecToString(message->payload());
+  std::unique_lock<std::mutex> lock(failMapMutex_);
+  auto it = failMessageCountMap_.find(key);
+  if (it == failMessageCountMap_.end()) {
+    failMessageCountMap_[key] = 0;
+  }
+  if (failMessageCountMap_[key] < numFailSends_) {
+    failMessageCountMap_[key]++;
+    lock.unlock();
+    auto jitFuture = c10::make_intrusive<JitFuture>(at::AnyClassType::get());
+    jitFuture->setError(std::make_exception_ptr(std::runtime_error(makeRPCError(
+        c10::str("Send attempt failed intentionally for ", key),
+        RPCErrorType::INTENTIONAL_FAILURE))));
+    return jitFuture;
+  } else {
+    lock.unlock();
+    return TensorPipeAgent::send(to, std::move(message), rpcTimeoutSeconds);
+  }
+}
+
+void FaultyTensorPipeAgent::pipeWrite(
+    const std::shared_ptr<tensorpipe::Pipe>& pipe,
+    c10::intrusive_ptr<Message> rpcMessage,
+    std::vector<c10::Device>&& devices,
+    std::vector<c10::Stream> streams,
+    std::function<void(const tensorpipe::Error&)> fn) noexcept {
+  float msgDelay = getDelayForMessage(rpcMessage->type());
+  if (msgDelay != 0) {
+    // Sleep for the specified delay for the message.
+    std::this_thread::sleep_for(std::chrono::milliseconds(
+        static_cast<int>(msgDelay * kSecToMsConversion)));
+  }
+  TensorPipeAgent::pipeWrite(pipe, rpcMessage, std::move(devices), streams, fn);
```

Comment:
Would I be correct if I assume the main difference is that we replaced `FaultyProcessGroupAgent::enqueueSend` with `FaultyTensorPipeAgent::pipeWrite`?

Reply:
Yes thats the main difference, `FaultyProcessGroupAgent::enqueueSend` as well as `FaultyProcessGroupAgent::sendToSelf`